### PR TITLE
Serialization using cereal

### PR DESCRIPTION
**Pull request by isuruf**
TODO:
- [x] rename `serialize.h` -> `serialize-cereal.h`
- [x] Add `loads`, `dumps` methods to `basic.h` and implement it using `cereal` in `basic.cpp`
- [x] Add tests
- [x] Make compiling conditional on cereal being available
- [x] Implement more classes.
- [x] ~Use something else than typeid. They are not stable across versions.~

### CODE REVIEW

File: symengine/serialize-cereal.h [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+// template <class Archive>
+// inline void save_basic(Archive &ar, const Rational &b)
+// {
+//     ar(b.get_num(), b.get_den());
+// }
+
+#define SYMENGINE_ENUM(type, Class)                     \
+    template <class Archive>                            \
+    inline void save_basic(Archive &ar, const Class &b) \
```
Comment (isuruf) [inline symengine/serialize-cereal.h:61]: You can use `save_basic(Archive &ar, const OneArgFunction &b)` and get rid of defining these multiple times.

File: symengine/serialize-cereal.h [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        return make_rcp<const Class>(arg1, arg2);                       \
+    }
+#include "symengine/type_codes_boolean_relational.inc"
+#undef SYMENGINE_ENUM
+
+
+template <class Archive, typename T>
+RCP<const Basic> load_basic(Archive &ar, RCP<const T> &b)
+{
+    const auto t_code = b->get_type_code();
```
Comment (isuruf) [inline symengine/serialize-cereal.h:175]: `b` is a dummy to allow writing code with inheritance. Don't use it.

**Comment by bjodah** [comment]
OK, I'm done for tonight (GMT+1 in Stockholm). If you get the chance to take a look: let me know what you think so far ☺.

File: symengine/basic.cpp [hunk 3]
```diff
@@ -24,6 +43,20 @@ std::string Basic::__str__() const
     return str(*this);
 }
 
+std::string Basic::dumps() const {
+    std::ostringstream oss;
+    cereal::BinaryOutputArchive{oss}(this->rcp_from_this());
```
Comment (isuruf) [inline symengine/basic.cpp:48]: Is `BinaryOutputArchive` portable?

File: symengine/serialize-cereal.h (hunk 2 above)
Comment (bjodah) [reply symengine/serialize-cereal.h:175]: Sure, but isn't it guaranteed that it inherits from Basic?

File: symengine/serialize-cereal.h (hunk 1 above)
Comment (bjodah) [reply symengine/serialize-cereal.h:61]: Ah, of course, will fix this.

File: symengine/serialize-cereal.h (hunk 2 above)
Comment (isuruf) [reply symengine/serialize-cereal.h:175]: Yes, but the pointer inside might be NULL, so you can't call methods on it.

File: symengine/basic.cpp (hunk 3 above)
Comment (bjodah) [reply symengine/basic.cpp:48]: Thanks, no it wasn't endianness is handled by PortableBinaryOutputArchive. Will change this.

File: symengine/serialize-cereal.h (hunk 1 above)
Comment (bjodah) [reply symengine/serialize-cereal.h]: So this really reduced the number of save_basic implementations. For `load_basic` I haven't yet come up with anything as elegant. I have the gut feeling it should be possible to do something leaner, any ideas?

**Comment by bjodah** [comment]
As for TypeIDs not being stable across versions of symengine: Cereal supports versioned serialization. We could write one new migrator for each new version of SymEngine, translating TypeIDs from version x to version x + 1. No performance impact for users keeping their serialized data up to date, and probably an acceptable cost for users using old data?

File: symengine/serialize-cereal.h (hunk 1 above)
Comment (isuruf) [reply symengine/serialize-cereal.h]: `load_basic(Archive &ar, RCP<const OneArgFunction> &b)` ?

Comment (bjodah) [reply symengine/serialize-cereal.h]: OneArgFunction is abstract, so I don't think that's possible?

Comment (bjodah) [reply symengine/serialize-cereal.h]: I mean I can dispatch on it, but then I'm at loss:
```c++
template <class Archive>
RCP<const Basic> load_basic(Archive &ar, RCP<const OneArgFunction> &)
{
    RCP<const Basic> arg;
    ar(arg);
    return basic_rcp_from_type_code_and_args(type_code, arg);
}
```
I guess I could change the signature of all `load_basic` dispatch functions to take `type_code` as an arguemnt, but then we still need the hypothetical `basic_rcp_from_type_code_and_args`?

Comment (isuruf) [reply symengine/serialize-cereal.h]: See https://github.com/symengine/symengine/pull/1704/commits/f7ada3705aa89797a764ec84b20a582b1e599ff7

**Comment by bjodah** [comment]
Ah, nice, `std::is_base_of` & `std::enable_if`, so this is SFINAE in action I presume? (I never seem to get it right whenever I try to use those).

**Comment by isuruf** [comment]
Yes, it is SFINAE

**Comment by bjodah** [comment]
Thanks, I got rid of those helpers I introduced before your SFINAE design.

So debugging SFINAE is a bit of a pain right? I'm trying to support URatPoly but it's not picking up my save_basic overload:
```
template <class Archive>
inline void save_basic(Archive &ar, const URatPoly &b)
{
    ar(b.get_var());
    ar(b.get_poly());
}
```
<details>
<summary>Excerpt from test_lambda_double</summary>
<pre>
Evaluate functions
-------------------------------------------------------------------------------
../symengine/tests/eval/test_lambda_double.cpp:187
...............................................................................

../symengine/tests/eval/test_lambda_double.cpp:187: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  ../symengine/serialize-cereal.h:23: void SymEngine::save_basic(Archive&,
  const SymEngine::Basic&) [with Archive = cereal::PortableBinaryOutputArchive]
  not supported: URatPoly (15), tan(x)*sec(x)
</pre>
</details>


I guess some one or several of my overloads for the building blocks here (`URatDict` -> `rational_class` -> `integer_class`) isn't quite working. But I don't know how to debug this behavior, any tips?

**Comment by isuruf** [comment]
Don't know how to debug the general case. In this case you probably need `Mul`

**Comment by bjodah** [comment]
Indeed, well spotted. You deduced this from "tan(x)*sec(x)"? (I never understood where the conversion to URatPoly occured)

**Comment by isuruf** [comment]
> You deduced this from "tan(x)*sec(x)"?

Yep. It was a type `Mul`. It wasn't a `URatPoly`. I fixed `type_code_name`.

**Comment by bjodah** [comment]
Of course there was a bug in that "helpful" function I wrote 😅

**Comment by bjodah** [comment]
Thank you for helping me here @isuruf, very much appreciated!

Alright, the test I added to `test_lambda_double` (which was exercising qutie a few types) is now passing.
I couldn't find any existing test case exercising the e.g. the creation of each of the types from `type_codes.inc` (I'm thinking that would be natural place to add a test of `dumps -> loads` to each and one of the types), did I miss any such test perhaps?
Otherwise I guess `test_serialize-cereal.cpp` is the right place for such an exhaustive test?

**Comment by rikardn** [comment]
I just poked in to resolve the merge conflict

**Comment by bjodah** [comment]
Thanks! I really want to see this finished at some point, but I have a thesis to write, so it might be a while.

**Comment by bocklund** [comment]
Is there any status update on this PR? Do the [tasks that @isuruf listed at the top](https://github.com/symengine/symengine/pull/1704#issue-737689185) still represent what needs to be done?

With this PR, SymEngine is in a state where it can completely replace SymPy for pycalphad ([PR](https://github.com/pycalphad/pycalphad/pull/376)) and ESPEI ([PR](https://github.com/PhasesResearchLab/ESPEI/pull/212)). My C++ is not that strong, but I would be happy to help if I can.

File: symengine/serialize-cereal.h [hunk 4]
```diff
@@ -0,0 +1,696 @@
+#pragma once
```
Comment (rikardn) [inline symengine/serialize-cereal.h:1]: This is non-standard. Use include guards instead?

**Review by rikardn — APPROVED** [review]
This is great work!

One question: What would happen for if attempting to serialize an object that isn't (yet) supported?

File: symengine/serialize-cereal.h (hunk 4 above)
Comment (bjodah) [reply symengine/serialize-cereal.h:1]: I'm curious: can you name a compiler which can compile symengine, but doesn't support `#pragma once`? 🙂

Comment (certik) [reply symengine/serialize-cereal.h:1]: I've seen the `#pragma once` used lately in C++. So far I always suggested to just use the old fashion include guards.

I am not against using `#pragma once`, it is definitely simpler. However I want to make sure we do not break the compilation.

@bjodah can you summarize the support in compilers for `#pragma once`? And why is it not standard?

Comment (isuruf) [reply symengine/serialize-cereal.h:1]: Good point. The only compiler I've seen is Oracle Development Studio which we have partial support for. https://github.com/symengine/symengine/issues/1646

Comment (bjodah) [reply symengine/serialize-cereal.h]: I don't know why it's not supported by the standard (the standard doesn't say much about the preprocessor from what I've seen). My question was genuine: I was curious about the existence of any compilers which lacks support for it (and I've seen multiple recommendations to prefer the pragma approach). But now we know: Oracle doesn't support it then. 🙂

**Comment by isuruf** [comment]
> One question: What would happen if attempting to serialize an object that isn't (yet) supported?

A `SymEngine::NotImplementedError` exception is thrown.

File: symengine/serialize-cereal.h (hunk 4 above)
Comment (certik) [reply symengine/serialize-cereal.h]: Do we want to support Oracle?

Comment (isuruf) [reply symengine/serialize-cereal.h]: It's not a big deal to keep using include guards instead of `pragma once`. I think we are using Oracle for R wrappers.

Comment (certik) [reply symengine/serialize-cereal.h]: Let's use the guards then.

**Comment by bjodah** [comment]
Do we want to add a `.gitattributes` file [so that vendored files](https://github.com/github/linguist/blob/master/docs/overrides.md#summary) from cereal doesn't show up in the "Files changed view"?, I think a `.gitattributes` file with an entry such as this would work:
```
symengine/utilities/cereal/include/cereal/* linguist-vendored
```

File: symengine/utilities/stream_fmt.h [hunk 5]
```diff
@@ -0,0 +1,21 @@
+#pragma once
```
Comment (bjodah) [inline symengine/utilities/stream_fmt.h:1]: I think I might have been guilty of this one.

File: LICENSE [hunk 6]
```diff
@@ -221,4 +221,33 @@ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 OTHER DEALINGS IN THE SOFTWARE.
+-------------------------------------------------------------------------------
+
+All files in symengine/utilities/catch were taken from the Trilinos package
```
Comment (bjodah) [inline LICENSE:226]: is the path really /catch for trilinos files?

**Review by bjodah — APPROVED** [review]
Great stuff! I only spotted two potential minor issues.

**Review by certik — APPROVED** [review]
